File: bench/bigmatrix/src/cdq.go
// Shape 2: cdq -- Go port of cdq.c (3D dominance, CDQ divide and conquer plus a
// weight Fenwick tree).
//
// Read n, then n triples (a b c) with values in [1,10^6].
//  1. sort by (a,b,c)   2. collapse duplicates, keeping cnt
//  3. CDQ over b, BIT over c: ans[j] += cnt[i] for every i<j that dominates j
//  4. f = ans + cnt (itself), reported answer = f - 1
//     print sum = SUM cnt_i * ans_i (u64 wrap) and max = max ans_i
package main

import (
	"fmt"
	"os"
	"sort"
)

// ---------------------------------------------------------------------------
// Fast stdin reader (see radix.go for the rationale; every Go shape carries its
// own copy because each file is built on its own).
// ---------------------------------------------------------------------------

const blockSize = 1 << 20

type scanner struct {
	buf []byte
	pos int
	end int
}

func newScanner() *scanner {
	return &scanner{buf: make([]byte, blockSize)}
}

func (s *scanner) fill() bool {
	n, _ := os.Stdin.Read(s.buf)
	s.pos, s.end = 0, n
	return n > 0
}

func (s *scanner) nextByte() int {
	if s.pos == s.end && !s.fill() {
		return -1
	}
	c := s.buf[s.pos]
	s.pos++
	return int(c)
}

// nextInt64 mirrors fastio.h's bm_i64.
func (s *scanner) nextInt64() int64 {
	c := s.nextByte()
	for c == ' ' || c == '\n' || c == '\t' || c == '\r' {
		c = s.nextByte()
	}
	neg := c == '-'
	if neg {
		c = s.nextByte()
	}
	var v int64
	for c >= '0' && c <= '9' {
		v = v*10 + int64(c-'0')
		c = s.nextByte()
	}
	if neg {
		return -v
	}
	return v
}

// nextUint64 mirrors fastio.h's bm_u64.
func (s *scanner) nextUint64() uint64 {
	for {
		for s.pos < s.end && (s.buf[s.pos] < '0' || s.buf[s.pos] > '9') {
			s.pos++
		}
		if s.pos != s.end {
			break
		}
		if !s.fill() {
			return 0
		}
	}

	var v uint64
	for {
		for s.pos < s.end {
			c := s.buf[s.pos]
			if c < '0' || c > '9' {
				s.pos++ // consume the separator
				return v
			}
			v = v*10 + uint64(c-'0')
			s.pos++
		}
		if !s.fill() {
			return v
		}
	}
}

// ---------------------------------------------------------------------------

// maxV is the coordinate bound.  The generator emits c in [1,maxV], so the
// Fenwick tree is indexed by c itself -- no +1 shift -- and stays maxV+1 wide
// exactly like the C reference's calloc.  bitAdd only ever touches indices in
// [1,maxV] (it stops as soon as i > maxV), so c = maxV lands on bit[maxV], the
// last element, and never one past the end.
const maxV = 1000000

// pt keeps the C reference's field widths on purpose.  ans accumulates in int32
// and is allowed to wrap; the final answer is computed in 64 bits, so the wrap
// has to happen at the same place C wraps it.
type pt struct {
	a, b, c int32
	cnt     int32
	ans     int32
}

// dominance is the whole CDQ state: the deduplicated points, the merge buffer,
// and the weight Fenwick tree over c.
type dominance struct {
	pts []pt
	buf []pt
	bit []int32
}

func (d *dominance) bitAdd(i int64, v int32) {
	for ; i <= maxV; i += i & -i {
		d.bit[i] += v
	}
}

func (d *dominance) bitSum(i int64) int32 {
	var s int32
	for ; i > 0; i -= i & -i {
		s += d.bit[i]
	}
	return s
}

// cdq counts, for every point, the weight of the points that dominate it.  It
// splits on index (a is already sorted), recurses, then sweeps the right half
// with the left half's contributions added to the tree in b order and undone
// afterwards; the merge at the end leaves both halves ordered by b.
func (d *dominance) cdq(l, r int64) {
	if l >= r {
		return
	}
	mid := l + (r-l)/2
	d.cdq(l, mid)
	d.cdq(mid+1, r)

	i, j := l, mid+1
	for ; j <= r; j++ {
		for i <= mid && d.pts[i].b <= d.pts[j].b {
			d.bitAdd(int64(d.pts[i].c), d.pts[i].cnt)
			i++
		}
		d.pts[j].ans += d.bitSum(int64(d.pts[j].c))
	}
	for t := l; t < i; t++ { // undo
		d.bitAdd(int64(d.pts[t].c), -d.pts[t].cnt)
	}

	i, j = l, mid+1
	k := l
	for i <= mid && j <= r {
		if d.pts[i].b <= d.pts[j].b {
			d.buf[k] = d.pts[i]
			i++
		} else {
			d.buf[k] = d.pts[j]
			j++
		}
		k++
	}
	for ; i <= mid; i++ {
		d.buf[k] = d.pts[i]
		k++
	}
	for ; j <= r; j++ {
		d.buf[k] = d.pts[j]
		k++
	}
	copy(d.pts[l:r+1], d.buf[l:r+1])
}

func main() {
	in := newScanner()
	n := int(in.nextInt64())

	raw := make([]pt, n)
	for i := range raw {
		raw[i] = pt{
			a:   int32(in.nextUint64()),
			b:   int32(in.nextUint64()),
			c:   int32(in.nextUint64()),
			cnt: 1,
		}
	}
	sort.Slice(raw, func(i, j int) bool {
		switch {
		case raw[i].a != raw[j].a:
			return raw[i].a < raw[j].a
		case raw[i].b != raw[j].b:
			return raw[i].b < raw[j].b
		default:
			return raw[i].c < raw[j].c
		}
	})

	// Collapse runs of equal triples, counting multiplicities.
	d := &dominance{
		pts: make([]pt, 0, n),
		buf: make([]pt, n),
		bit: make([]int32, maxV+1),
	}
	for i := 0; i < n; {
		j := i + 1
		for j < n && raw[j].a == raw[i].a && raw[j].b == raw[i].b && raw[j].c == raw[i].c {
			j++
		}
		raw[i].cnt = int32(j - i)
		raw[i].ans = 0
		d.pts = append(d.pts, raw[i])
		i = j
	}
	d.buf = d.buf[:len(d.pts)]

	d.cdq(0, int64(len(d.pts))-1)

	var sum uint64
	var mx int64
	for i := range d.pts {
		ans := int64(d.pts[i].ans) + int64(d.pts[i].cnt) - 1
		sum += uint64(d.pts[i].cnt) * uint64(ans)
		if ans > mx {
			mx = ans
		}
	}
	fmt.Printf("sum=%d max=%d\n", sum, mx)
}

File: bench/bigmatrix/src/cdq.rs
// Shape 2: cdq -- Rust port of cdq.c (3D dominance, CDQ + weight BIT).
//
// Read n, then n triples (a b c) with values in [1,10^6].
//   1. sort by (a,b,c)   2. collapse duplicates, keeping cnt
//   3. CDQ over b, BIT over c: ans[j] += cnt[i] for every i<j that dominates j
//   4. f = ans + cnt (itself), reported answer = f - 1
//   print sum = SUM cnt_i * ans_i (u64 wrap) and max = max ans_i

use std::io::{self, Read};

/* ---------- fast stdin: a port of fastio.h ---------- */

/// 1 MiB read buffer plus hand-written integer parsing.  `BufRead::lines()`
/// followed by `str::parse` would dominate a run that feeds 10^7 numbers.
struct FastIn<R: Read> {
    r: R,
    buf: Box<[u8]>,
    len: usize,
    pos: usize,
}

impl<R: Read> FastIn<R> {
    fn new(r: R) -> Self {
        FastIn {
            r,
            buf: vec![0u8; 1 << 20].into_boxed_slice(),
            len: 0,
            pos: 0,
        }
    }

    /// One byte, or -1 at end of input (bm_get).
    #[inline]
    fn get(&mut self) -> i32 {
        if self.pos >= self.len {
            self.len = self.r.read(&mut self.buf).unwrap_or(0);
            self.pos = 0;
            if self.len == 0 {
                return -1;
            }
        }
        let c = self.buf[self.pos] as i32;
        self.pos += 1;
        c
    }

    /// Signed integer, skipping ' ', '\n', '\t', '\r' (bm_i64).
    #[inline]
    fn i64(&mut self) -> i64 {
        let mut c = self.get();
        let mut neg = false;
        while c == b' ' as i32 || c == b'\n' as i32 || c == b'\t' as i32 || c == b'\r' as i32 {
            c = self.get();
        }
        if c == b'-' as i32 {
            neg = true;
            c = self.get();
        }
        let mut v: i64 = 0;
        while is_digit(c) {
            v = v.wrapping_mul(10).wrapping_add((c - b'0' as i32) as i64);
            c = self.get();
        }
        if neg {
            -v
        } else {
            v
        }
    }

    /// Unsigned integer, skipping any non-digit (bm_u64).
    #[inline]
    fn u64(&mut self) -> u64 {
        let mut c = self.get();
        let mut v: u64 = 0;
        while !is_digit(c) {
            if c < 0 {
                return 0;
            }
            c = self.get();
        }
        while is_digit(c) {
            v = v.wrapping_mul(10).wrapping_add((c - b'0' as i32) as u64);
            c = self.get();
        }
        v
    }
}

#[inline]
fn is_digit(c: i32) -> bool {
    (b'0' as i32..=b'9' as i32).contains(&c)
}

/* ---------- shape ---------- */

const V: usize = 1_000_000;

/// 5 x i32 = 20 bytes, the same point record layout as the C reference.
#[derive(Clone, Copy, Default)]
struct Pt {
    a: i32,
    b: i32,
    c: i32,
    cnt: i32,
    ans: i32,
}

/// Weight BIT over c, 1-based: c is in [1, V], the array is V+1 wide, and the
/// loop keeps `i <= V`, so every access is in bounds (c = 10^6 -> index V).
#[inline]
fn bit_add(bit: &mut [i32], mut i: usize, v: i32) {
    while i <= V {
        bit[i] += v;
        i += i & i.wrapping_neg();
    }
}

#[inline]
fn bit_sum(bit: &[i32], mut i: usize) -> i32 {
    let mut s = 0i32;
    while i > 0 {
        s += bit[i];
        i -= i & i.wrapping_neg();
    }
    s
}

struct Cdq {
    p: Vec<Pt>,
    tmp: Vec<Pt>,
    bit: Vec<i32>,
}

impl Cdq {
    /// CDQ over b with a weight BIT over c.  Indices stay i64 so that the
    /// empty range of an empty point set (l = 0, r = -1) behaves as in C.
    fn solve(&mut self, l: i64, r: i64) {
        if l >= r {
            return;
        }
        let mid = l + (r - l) / 2;
        self.solve(l, mid);
        self.solve(mid + 1, r);

        let mut i = l;
        let mut j = mid + 1;
        while j <= r {
            while i <= mid && self.p[i as usize].b <= self.p[j as usize].b {
                let (c, cnt) = (self.p[i as usize].c, self.p[i as usize].cnt);
                bit_add(&mut self.bit, c as usize, cnt);
                i += 1;
            }
            let c = self.p[j as usize].c;
            self.p[j as usize].ans += bit_sum(&self.bit, c as usize);
            j += 1;
        }
        for t in l..i {
            let (c, cnt) = (self.p[t as usize].c, self.p[t as usize].cnt);
            bit_add(&mut self.bit, c as usize, -cnt); // undo
        }

        let mut i = l;
        let mut j = mid + 1;
        let mut k = l;
        while i <= mid && j <= r {
            if self.p[i as usize].b <= self.p[j as usize].b {
                self.tmp[k as usize] = self.p[i as usize];
                i += 1;
            } else {
                self.tmp[k as usize] = self.p[j as usize];
                j += 1;
            }
            k += 1;
        }
        while i <= mid {
            self.tmp[k as usize] = self.p[i as usize];
            i += 1;
            k += 1;
        }
        while j <= r {
            self.tmp[k as usize] = self.p[j as usize];
            j += 1;
            k += 1;
        }
        let (l, r) = (l as usize, r as usize);
        self.p[l..=r].copy_from_slice(&self.tmp[l..=r]);
    }
}

fn main() {
    let mut r = FastIn::new(io::stdin().lock());
    let n = r.i64() as usize;

    let mut raw: Vec<Pt> = (0..n)
        .map(|_| Pt {
            a: r.u64() as i32,
            b: r.u64() as i32,
            c: r.u64() as i32,
            cnt: 1,
            ans: 0,
        })
        .collect();

    raw.sort_unstable_by(|x, y| (x.a, x.b, x.c).cmp(&(y.a, y.b, y.c)));

    // Collapse duplicates; the slice is sorted, so equals are adjacent.
    let mut p: Vec<Pt> = Vec::new();
    let mut i = 0usize;
    while i < n {
        let q = raw[i];
        let j = i
            + raw[i..]
                .iter()
                .take_while(|z| z.a == q.a && z.b == q.b && z.c == q.c)
                .count();
        p.push(Pt {
            cnt: (j - i) as i32,
            ..q
        });
        i = j;
    }
    let m = p.len() as i64;

    let mut solver = Cdq {
        p,
        // Zero-initialised on purpose: Rust (like extC) guarantees initialised
        // scratch, so this ~40 MB memset is a deliberate language-level cost
        // and should not be "optimised" away with an unsafe set_len.
        tmp: vec![Pt::default(); m as usize],
        bit: vec![0i32; V + 1],
    };
    solver.solve(0, m - 1);

    let mut sum: u64 = 0;
    let mut mx: i64 = 0;
    for q in &solver.p {
        let ans = q.ans as i64 + q.cnt as i64 - 1;
        sum = sum.wrapping_add((q.cnt as u64).wrapping_mul(ans as u64));
        if ans > mx {
            mx = ans;
        }
    }
    println!("sum={} max={}", sum, mx);
}
